src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -31,21 +31,20 @@ public static class TagHelperOutputExtensions
             [NotNull] string attributeName,
             [NotNull] TagHelperContext context)
         {
-            if (!tagHelperOutput.Attributes.ContainsKey(attributeName))
+            if (!tagHelperOutput.Attributes.ContainsName(attributeName))
             {
+                IReadOnlyTagHelperAttribute entry;
+
                 // We look for the original attribute so we can restore the exact attribute name the user typed.
                 // Approach also ignores changes made to tagHelperOutput[attributeName].
-                var entry = context.AllAttributes.FirstOrDefault(
-                    attribute => attribute.Key.Equals(attributeName, StringComparison.OrdinalIgnoreCase));
-
-                if (entry.Equals(default(KeyValuePair<string, object>)))
+                if (!context.AllAttributes.TryGetAttribute(attributeName, out entry))
                 {
                     throw new ArgumentException(
                         Resources.FormatTagHelperOutput_AttributeDoesNotExist(attributeName, nameof(TagHelperContext)),
                         nameof(attributeName));
                 }
-                
-                tagHelperOutput.Attributes.Add(entry.Key, entry.Value);
+
+                tagHelperOutput.Attributes.Add(entry.Name, entry.Value);
```

Comment:
this method should restore all attributes with given name.  aspnet/Razor#279 is incomplete as long as this method only handles the first match.